tests/runtimes/substantial/common.ts

```diff
@@ -0,0 +1,322 @@
+              runIds.push(...[one, two, three]);
+            })
+            .on(e);
+        }
+      );
+
+      // let's wait for a bit to make sure interrupts are doing their jobs
+      await sleep(5 * 1000);
+
+      await t.should(
+        `fire events for ${emails.join(", ")} (${backendName})`,
+        async () => {
+          await gql`
+            mutation {
+              # will pass
+              one: send_confirmation(
+                run_id: $one_run_id
+                event: { payload: true }
+              )
+              # will throw
+              two: send_confirmation(
+                run_id: $two_run_id
+                event: { payload: false }
+              )
+              # will abort
+              three: abort_email_confirmation(run_id: $three_run_id)
+            }
+          `
+            .withVars({
+              one_run_id: runIds[0],
+              two_run_id: runIds[1],
+              three_run_id: runIds[2],
+            })
+            .expectData({
+              one: runIds[0],
+              two: runIds[1],
+              three: runIds[2],
+            })
+            .on(e);
+        }
+      );
+
+      // This is arbitrary, if ops are leaking that means it should be increased
+      // Noticing the fired event may take a few seconds depending on the interrupt relaunch time and poll interval
+      // Once noticed the workflow will complete and the worker destroyed
+      // About ~10s for fs and memory
+      // About ~12, 16s local redis
+      await sleep(delays.awaitEmailCompleteSec * 1000);
+
+      await t.should(`complete execution (${backendName})`, async () => {
+        await gql`
+          query {
+            email_results {
+              ongoing {
+                count
+              }
+              completed {
+                count
+                runs {
+                  run_id
+                  result {
+                    status
+                    value
+                  }
+                }
+              }
+            }
+          }
+        `
+          .expectBody((body) => {
+            assertEquals(
+              body?.data?.email_results?.ongoing?.count,
+              0,
+              "0 workflow currently running"
+            );
+
+            assertEquals(
+              body?.data?.email_results?.completed?.count,
+              3,
+              "3 workflows completed"
+            );
+
+            const localSorter = (a: any, b: any) =>
+              a.run_id.localeCompare(b.run_id);
+
+            const received =
+              body?.data?.email_results?.completed?.runs ?? ([] as Array<any>);
+            const expected = [
+              {
+                result: {
+                  status: "COMPLETED",
+                  value: 'Email sent to one@example.com: "confirmed!"',
+                },
+                run_id: runIds[0],
+              },
+              {
+                result: {
+                  status: "COMPLETED_WITH_ERROR",
```

Comment:
Are retry policies and stuff in scope for substantial?

Reply:
Yes, something like `ctx.save(() => ..., { backoff: Time.seconds(10), retries: 10 })`.